tensorflow/core/kernels/softmax_op_gpu.cu.cc

```diff
@@ -72,28 +72,94 @@ struct softmax_traits<Eigen::half> {
   using accumulator_type = float;
 };
 
-template <typename T, typename U>
+template <typename T, typename U, int kUnroll>
 __global__ void GenerateNormalizedProb(const T* logits, const U* sum_probs,
                                        const T* max_logits, T* output,
                                        const int num_rows, const int num_cols,
                                        const bool in_log_space) {
-  const int tid = blockIdx.x * blockDim.x + threadIdx.x;
-
-  const int row = tid / num_cols;
-  const int col = tid % num_cols;
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  int row, col;
 
   // TODO(jamesqin): change to half2 load when inputs are Eigen::half.
-  U input = strict_cast<U>(logits[tid]);
-  U max_val = strict_cast<U>(ldg(max_logits + row));
-  U result;
-
-  if (row < num_rows && col < num_cols) {
-    if (in_log_space) {
-      result = input - max_val - log(ldg(sum_probs + row));
-    } else {
-      result = exp(input - max_val) / ldg(sum_probs + row);
+  U input[kUnroll];
+  U max_val[kUnroll];
+  U result[kUnroll];
+  for (int i = 0; i < kUnroll; i++) {
+    row = tid / num_cols;
+    col = tid % num_cols;
+    if (row < num_rows && col < num_cols) {
+      input[i] = strict_cast<U>(logits[tid]);
+      max_val[i] = strict_cast<U>(ldg(max_logits + row));
+    }
+    tid += gridDim.x * blockDim.x;
+  }
+
+  tid = blockIdx.x * blockDim.x + threadIdx.x;
+  for (int i = 0; i < kUnroll; i++) {
+    row = tid / num_cols;
+    col = tid % num_cols;
+    if (row < num_rows && col < num_cols) {
+      if (in_log_space) {
+        result[i] = input[i] - max_val[i] - log(ldg(sum_probs + row));
+      } else {
+        result[i] = exp(input[i] - max_val[i]) / ldg(sum_probs + row);
+      }
+      output[tid] = strict_cast<T>(result[i]);
+    }
+    tid += gridDim.x * blockDim.x;
+  }
+}
+
+template <>
+__global__ void GenerateNormalizedProb<Eigen::half, float, 8>(
+    const Eigen::half* logits, const float* sum_probs,
+    const Eigen::half* max_logits, Eigen::half* output,
+    const int num_rows, const int num_cols, const bool in_log_space) {
+  const int kUnroll = 8;
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  int idx[kUnroll];
+  int row[kUnroll];
+
+  float input[kUnroll];
+  float max_val[kUnroll];
+  float result[kUnroll];
+
+  if (tid * kUnroll + kUnroll - 1 < num_rows * num_cols) {
+    ulonglong2 logits_d =
+        *reinterpret_cast<const ulonglong2*>(logits + tid * kUnroll);
+    Eigen::half* logits_h = reinterpret_cast<Eigen::half*>(&logits_d);
+    ulonglong2 output_d;
+    Eigen::half* output_h = reinterpret_cast<Eigen::half*>(&output_d);
+
+    for (int i = 0; i < kUnroll; i++) {
```

Comment:
Would it be better to make logits_d a local variable and explicitly load kUnroll logits via logits_d instead of accessing logits_h[i] in the loop (like you do for the writes)? Or does  the compiler figure out how to coalesce those reads automatically?

Reply:
Thanks for your comments.
I think we already load the kUnroll logits via logits_d (line 129). What inside the loop (line 134) is accessing each half value (via logits_h[i]) to do the actual computation.